resources/type-system/inference.md

```diff
@@ -1139,181 +1204,152 @@ as follows:
-  or extension type, or the "on" type of the immediately enclosing extension.
+Expression inference of `this`, regardless of context, produces the elaborated
+expression `this`.
 
-- If there is no immediately enclosing class, enum, mixin, extension type, or
-  extension, then there is a compile-time error.
+_The static type of `this` is the interface type of the immediately enclosing
+class, enum, mixin, or extension type, or the "on" type of the immediately
+enclosing extension. The runtime behavior of `this` is to evaluate to the target
+of the current instance member invocation, which is guaranteed to be an instance
+of this type. So soundness is satisfied._
 
-- If the expression `this` appears inside a factory constructor, a constructor's
-  initializer list, a static method or variable initializer, or in the
-  initializing expression of a non-late instance variable, then there is a
-  compile-time error.
+### Logical boolean expressions
 
-- Let `m` be an expression artifact whose runtime behavior is to complete with a
-  value that is the target of the current instance member invocation.
+Expression inference of a logical "and" expression (`e_1 && e_2`) or a logical
+"or" expression (`e_1 || e_2`), regardless of context, produces an elaborated
+expression `m` with static type `bool`, where `m` is determined as follows:
 
-_Expression soundness follows from the fact that within a class, enum, mixin, or
-extension type having interface type `T`, the target of any instance member
-invocation is always an instance of `T`, and within an extension, the target of
-any instance member invocation is always an instance of the extension's "on"
-type._
-
-## Logical boolean expressions
-
-Expression type inference of a logical "and" expression (`e_1 && e_2`) or a
-logical "or" expression (`e_1 || e_2`) produces an expression artifact `m` with
-static type `bool`, where `m` is determined as follows:
-
-- Let `m_1` be the result of performing expression type inference on `e_1`, in
+- Let `m_1` be the result of performing expression inference on `e_1`, in
   context `bool`, and then coercing the result to type `bool`.
 
-- Let `m_2` be the result of performing expression type inference on `e_2`, in
+- Let `m_2` be the result of performing expression inference on `e_2`, in
   context `bool`, and then coercing the result to type `bool`.
 
-- Let `m` be an expression artifact whose runtime behavior is as follows:
+- _It follows, from the soundness invariant of coercions, that the static type
+  of `m_1` and `m_2` are both guaranteed to be a subtype of `bool`._
 
-  - Execute expression artifact `m_1`, and let `o_1` be the resulting value. _By
-    expression soundness, `o_1` will be an instance of the type `bool`._
+- If `e` is of the form `e_1 && e_2`, let `m` be `m_1 && m_2`. _It is valid to
+  form this elaborated expression because the static type of `m_1` and `m_2` are
+  guaranteed to be a subtype of `bool`._
```

Comment:
(I'd say to coerce that to the context type, but we don't allow implicit tearoff of extension `call` methods, right?)

(Complete side track: should we let the static type of the expression be **UP** of the static types of the `m_1` and `m_2` elaborated expressions? If both are the same extension type implementing `bool`, we could let the result have that type too!)

Reply:
You’re right, we don’t do implicit tearoff of extension call methods, so there’s no coercion that would change `bool` to any other type.

Also, since my goal here is to be descriptive of what the implementations currently do, I don’t want to specify that we coerce to the context "type" because the implementations *don’t* try to coerce logical boolean expressions to the context. (Indeed, they never really try to coerce to the context per se; it just looks like they do because coercions always happen at assignability check points, and when there's an assignability check, the context for the subexpression is _usually_ the same as the type used in the assignability check).

As to your side-track, that’s a really interesting thought. I’m torn between a desire to implement an interesting language and a desire to write fewer crazy tests 😃